node/client/src/lib.rs

```diff
@@ -169,37 +216,38 @@ impl Client {
             }
         }
 
-        if self.beacon_chain.chain.is_known(&block.hash) {
-            info!(target: "client", "The block was already imported, before we managed to produce it.");
-            io::stdout().flush().expect("Could not flush stdout");
-            None
-        } else {
-            self.shard_chain.insert_block(&shard_block.clone(), transaction, tx_results, new_receipts);
-            self.beacon_chain.chain.insert_block(block.clone());
-            info!(target: "client",
+        assert!(
+            !self.beacon_chain.chain.is_known(&block.hash),
+            "The block was already imported, before we managed to produce it.\
+             This should never happen, because block production is atomic."
+        );
+
+        self.shard_chain.insert_block(&shard_block.clone(), transaction, tx_results, new_receipts);
+        self.beacon_chain.chain.insert_block(block.clone());
+        info!(target: "client",
                   "Producing block index: {:?}, beacon = {:?}, shard = {:?}",
                   block.body.header.index, block.hash, shard_block.hash);
-            io::stdout().flush().expect("Could not flush stdout");
-            // Just produced blocks should be the best in the blockchain.
-            assert_eq!(self.shard_chain.chain.best_block().hash, shard_block.hash);
-            assert_eq!(self.beacon_chain.chain.best_block().hash, block.hash);
-            // Update the authority.
-            self.update_authority(&block.header());
-            Some((block, shard_block))
-        }
+        io::stdout().flush().expect("Could not flush stdout");
```

Comment:
Kind of strange place for flushing stdout. Shouldn't log do it already?

Reply:
Though I haven't verified it, I think it does not. I added it long time ago: I was running devnet and submitting transactions to it, but it did not print the block production logs, so I thought it was broken until I added the flush and it started printing them. The working hypothesis is that it does not.